File: lib/core/services/maps_web_service.dart
import 'dart:async';
import 'dart:js' as js;
import 'package:flutter/foundation.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';

/// Web-specific Google Maps service using JS API (avoids CORS)
/// This bridges to maps_web_service.js which uses the Google Maps JS API
class MapsWebService {
  static bool get isSupported => kIsWeb;

  /// Get autocomplete suggestions
  static Future<List<Map<String, String>>> getSuggestions(String input) async {
    if (!isSupported) return [];

    final completer = Completer<List<Map<String, String>>>();

    try {
      final service = js.context['MapsWebService'];

      if (service == null) {
        debugPrint('[MapsWebService] JS service not found');
        return [];
      }

      // Create a JS callback function
      final callback = js.allowInterop((dynamic results) {
        final suggestions = <Map<String, String>>[];
        if (results is List) {
          for (var item in results) {
            // Convert JsObject to Dart Map
            final jsItem = js.JsObject.fromBrowserObject(item as js.JsObject);
            final desc = jsItem['description'] as String?;
            final placeId = jsItem['placeId'] as String?;
            if (desc != null && placeId != null) {
              suggestions.add({'description': desc, 'placeId': placeId});
            }
          }
        }
        if (!completer.isCompleted) {
          completer.complete(suggestions);
        }
      });

      service.callMethod('getSuggestions', [input, callback]);

      // Timeout after 2 seconds
      Future.delayed(const Duration(seconds: 2), () {
        if (!completer.isCompleted) {
          completer.complete([]);
        }
      });
    } catch (e) {
      debugPrint('[MapsWebService] Error: $e');
      return [];
    }

    return completer.future;
  }

  /// Get place details (lat/lng)
  static Future<LatLng?> getPlaceDetails(String placeId) async {
    if (!isSupported) return null;

    final completer = Completer<LatLng?>();

    try {
      final service = js.context['MapsWebService'];

      if (service == null) return null;

      final callback = js.allowInterop((dynamic result) {
        if (result != null) {
          final jsResult = js.JsObject.fromBrowserObject(result as js.JsObject);
          final lat = jsResult['lat'] as num?;
          final lng = jsResult['lng'] as num?;
          if (lat != null && lng != null) {
            completer.complete(LatLng(lat.toDouble(), lng.toDouble()));
            return;
          }
        }
        completer.complete(null);
      });

      service.callMethod('getPlaceDetails', [placeId, callback]);

      Future.delayed(const Duration(seconds: 2), () {
        if (!completer.isCompleted) completer.complete(null);
      });
    } catch (e) {
      debugPrint('[MapsWebService] Place details error: $e');
      return null;
    }

    return completer.future;
  }

  /// Get directions
  static Future<Map<String, dynamic>?> getDirections(
    LatLng origin,
    LatLng destination,
  ) async {
    if (!isSupported) return null;

    final completer = Completer<Map<String, dynamic>?>();

    try {
      final service = js.context['MapsWebService'];

      if (service == null) return null;

      final callback = js.allowInterop((dynamic result) {
        if (result != null) {
          final jsResult = js.JsObject.fromBrowserObject(result as js.JsObject);
          final distanceText = jsResult['distanceText'] as String?;
          final distanceValue = jsResult['distanceValue'] as num?;
          final durationText = jsResult['durationText'] as String?;
          final pointsList = jsResult['points'];

          if (distanceText != null &&
              pointsList != null &&
              pointsList is List) {
            final points = <LatLng>[];
            for (var p in pointsList) {
              final point = js.JsObject.fromBrowserObject(p as js.JsObject);
              final lat = point['lat'] as num?;
              final lng = point['lng'] as num?;
              if (lat != null && lng != null) {
                points.add(LatLng(lat.toDouble(), lng.toDouble()));
              }
            }

            completer.complete({
              'distanceText': distanceText,
              'distanceValue': distanceValue?.toDouble() ?? 0.0,
              'durationText': durationText ?? '',
              'points': points,
            });
            return;
          }
        }
        completer.complete(null);
      });

      service.callMethod('getDirections', [
        origin.latitude,
        origin.longitude,
        destination.latitude,
        destination.longitude,
        callback,
      ]);

      Future.delayed(const Duration(seconds: 3), () {
        if (!completer.isCompleted) completer.complete(null);
      });
    } catch (e) {
      debugPrint('[MapsWebService] Directions error: $e');
      return null;
    }

    return completer.future;
  }

  /// Reverse geocode lat/lng to address
  static Future<Map<String, String>?> reverseGeocode(LatLng location) async {
    if (!isSupported) return null;

    final completer = Completer<Map<String, String>?>();

    try {
      final service = js.context['MapsWebService'];

      if (service == null) return null;

      final callback = js.allowInterop((dynamic result) {
        if (result != null) {
          final jsResult = js.JsObject.fromBrowserObject(result as js.JsObject);
          final address = jsResult['address'] as String?;
          final city = jsResult['city'] as String?;
          final fullAddress = jsResult['fullAddress'] as String?;

          if (address != null && city != null) {
            completer.complete({
              'address': address,
              'city': city,
              'fullAddress': fullAddress ?? address,
            });
            return;
          }
        }
        completer.complete(null);
      });

      service.callMethod('reverseGeocode', [
        location.latitude,
        location.longitude,
        callback,
      ]);

      Future.delayed(const Duration(seconds: 2), () {
        if (!completer.isCompleted) completer.complete(null);
      });
    } catch (e) {
      debugPrint('[MapsWebService] Reverse geocode error: $e');
      return null;
    }

    return completer.future;
  }
}

File: web/maps_web_service.js
// Google Maps Web Service - Called from Flutter via JS interop
window.MapsWebService = {
  autocompleteService: null,
  placesService: null,
  directionsService: null,

  init: function() {
    if (window.google && window.google.maps) {
      this.autocompleteService = new window.google.maps.places.AutocompleteService();
      this.directionsService = new window.google.maps.DirectionsService();
      console.log('[MapsWebService] Initialized');
    }
  },

  // Get autocomplete suggestions
  getSuggestions: function(input, callback) {
    if (!this.autocompleteService) {
      this.init();
    }
    
    if (!this.autocompleteService) {
      console.error('[MapsWebService] AutocompleteService not available');
      callback([]);
      return;
    }

    this.autocompleteService.getPlacePredictions(
      { input: input, types: ['geocode'] },
      function(predictions, status) {
        if (status === 'OK' || status === 'ZERO_RESULTS') {
          const results = predictions ? predictions.map(function(p) {
            return {
              description: p.description,
              placeId: p.place_id,
              mainText: p.structured_formatting ? p.structured_formatting.main_text : p.description
            };
          }) : [];
          console.log('[MapsWebService] Got ' + results.length + ' suggestions');
          callback(results);
        } else {
          console.error('[MapsWebService] Autocomplete error:', status);
          callback([]);
        }
      }
    );
  },

  // Get place details (lat/lng)
  getPlaceDetails: function(placeId, callback) {
    if (!window.google || !window.google.maps) {
      callback(null);
      return;
    }

    // Create a dummy map element for PlacesService
    const dummyDiv = document.createElement('div');
    const map = new window.google.maps.Map(dummyDiv, {
      center: { lat: 0, lng: 0 },
      zoom: 1
    });
    
    const service = new window.google.maps.places.PlacesService(map);
    
    service.getDetails(
      { placeId: placeId, fields: ['geometry'] },
      function(place, status) {
        if (status === 'OK' && place.geometry) {
          callback({
            lat: place.geometry.location.lat(),
            lng: place.geometry.location.lng()
          });
        } else {
          console.error('[MapsWebService] Place details error:', status);
          callback(null);
        }
      }
    );
  },

  // Reverse geocode (lat/lng to address)
  reverseGeocode: function(lat, lng, callback) {
    if (!window.google || !window.google.maps) {
      callback(null);
      return;
    }

    const geocoder = new window.google.maps.Geocoder();
    
    geocoder.geocode(
      { location: { lat: lat, lng: lng } },
      function(results, status) {
        if (status === 'OK' && results && results.length > 0) {
          const result = results[0];
          const address = result.formatted_address;
          
          // Extract components
          let city = '';
          let subLocality = '';
          for (let i = 0; i < result.address_components.length; i++) {
            const component = result.address_components[i];
            if (component.types.includes('locality')) {
              city = component.long_name;
            }
            if (component.types.includes('sublocality') || component.types.includes('sublocality_level_1')) {
              subLocality = component.long_name;
            }
          }
          
          callback({
            address: address,
            city: subLocality || city || 'Unknown',
            fullAddress: address
          });
        } else {
          console.error('[MapsWebService] Geocoding error:', status);
          callback(null);
        }
      }
    );
  },

  // Get directions
  getDirections: function(originLat, originLng, destLat, destLng, callback) {
    if (!this.directionsService) {
      this.init();
    }

    if (!this.directionsService) {
      callback(null);
      return;
    }

    this.directionsService.route(
      {
        origin: { lat: originLat, lng: originLng },
        destination: { lat: destLat, lng: destLng },
        travelMode: 'DRIVING'
      },
      function(response, status) {
        if (status === 'OK') {
          const route = response.routes[0];
          const leg = route.legs[0];
          
          // Extract polyline points
          const points = [];
          const path = route.overview_path;
          for (let i = 0; i < path.length; i++) {
            points.push({
              lat: path[i].lat(),
              lng: path[i].lng()
            });
          }

          callback({
            distanceText: leg.distance.text,
            distanceValue: leg.distance.value,
            durationText: leg.duration.text,
            points: points
          });
        } else {
          console.error('[MapsWebService] Directions error:', status);
          callback(null);
        }
      }
    );
  }
};

// Initialize when Google Maps is ready
if (window.google && window.google.maps) {
  window.MapsWebService.init();
} else {
  // Wait for Google Maps to load
  window.addEventListener('load', function() {
    setTimeout(function() {
      window.MapsWebService.init();
    }, 1000);
  });
}
